samples/brb-channel/control-module.go

```diff
@@ -0,0 +1,133 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"github.com/filecoin-project/mir/pkg/events"
+	"github.com/filecoin-project/mir/pkg/modules"
+	"github.com/filecoin-project/mir/pkg/pb/brbctpb"
+	"github.com/filecoin-project/mir/pkg/pb/brbdxrpb"
+	"github.com/filecoin-project/mir/pkg/pb/brbpb"
+	"github.com/filecoin-project/mir/pkg/pb/eventpb"
+	"time"
+)
+
+type controlModule struct {
+	eventsOut                 chan *events.EventList
+	lastId                    int64
+	sentMessages              int64
+	isLeader                  bool
+	currentBenchmark          *Benchmark
+	broadcastRequestGenerator func(int64, *[]byte, string) *events.EventList
+	broadcastDeliverValidator func([]byte)
+	tests                     []*Benchmark
+	lastTest                  int
+}
+
+func newControlModule(isLeader bool, broadcastRequestGenerator func(int64, *[]byte, string) *events.EventList, broadcastDeliverValidator func([]byte), tests []*Benchmark) modules.ActiveModule {
+	return &controlModule{
+		eventsOut:                 make(chan *events.EventList),
+		lastId:                    0,
+		isLeader:                  isLeader,
+		broadcastRequestGenerator: broadcastRequestGenerator,
+		broadcastDeliverValidator: broadcastDeliverValidator,
+		tests:                     tests,
+		lastTest:                  0,
+	}
+}
+
+func (m *controlModule) ImplementsModule() {}
+
+func (m *controlModule) ApplyEvents(ctx context.Context, events *events.EventList) error {
+	iter := events.Iterator()
+	for event := iter.Next(); event != nil; event = iter.Next() {
+		switch event.Type.(type) {
+
+		case *eventpb.Event_Init:
+			m.newIteration()
+
+		case *eventpb.Event_Brbct:
+			brbEvent := event.Type.(*eventpb.Event_Brbct).Brbct
+			switch brbEvent.Type.(type) {
+
+			case *brbctpb.Event_Deliver:
+				deliverEvent := brbEvent.Type.(*brbctpb.Event_Deliver).Deliver
+				m.broadcastDeliverValidator(deliverEvent.Data)
+				m.newIteration()
+			default:
+				return fmt.Errorf("unknown brb event type: %T", brbEvent.Type)
+			}
+
+		case *eventpb.Event_Brbdxr:
+			brbEvent := event.Type.(*eventpb.Event_Brbdxr).Brbdxr
+			switch brbEvent.Type.(type) {
+
+			case *brbdxrpb.Event_Deliver:
+				deliverEvent := brbEvent.Type.(*brbdxrpb.Event_Deliver).Deliver
+				m.broadcastDeliverValidator(deliverEvent.Data)
+				m.newIteration()
+			default:
+				return fmt.Errorf("unknown brb event type: %T", brbEvent.Type)
+			}
+
+		case *eventpb.Event_Brb:
+			brbEvent := event.Type.(*eventpb.Event_Brb).Brb
+			switch brbEvent.Type.(type) {
+
+			case *brbpb.Event_Deliver:
+				deliverEvent := brbEvent.Type.(*brbpb.Event_Deliver).Deliver
+				m.broadcastDeliverValidator(deliverEvent.Data)
+				m.newIteration()
+			default:
+				return fmt.Errorf("unknown brb event type: %T", brbEvent.Type)
+			}
```

Comment:
This switch-case is unnecessary, just use the same events for all 3 implementations.